components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +291,234 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
+     *
+     * @param clientId     Client ID of the application.
+     * @return Registered client authentication method for the application.
+     * @throws OAuthClientAuthnException
+     */
+    public String retrieveRegisteredAuthMethod(String clientId) throws OAuthClientAuthnException {
+
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (StringUtils.isEmpty(serviceProvider.getCertificateContent())) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (OAuthConstants.TOKEN_ENDPOINT_AUTH_METHOD.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return OAuthConstants.NOT_APPLICABLE;
+    }
+
+    /**
+     * Check whether the application should support FAPI.
+     *
+     * @param clientId       Client ID of the application.
+     * @return Whether the application should support FAPI.
+     * @throws OAuthClientAuthnException
+     */
+    private String isFapiApplication(String clientId) throws OAuthClientAuthnException {
+
+        String isFapiApp = "IsFAPIApp";
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (StringUtils.isEmpty(serviceProvider.getCertificateContent())) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (isFapiApp.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return null;
+    }
+
+    /**
+     * Obtain the client ID of the application from the request.
+     *
+     * @param request       Http servlet request.
+     * @return Client ID of the application.
+     * @throws OAuthClientAuthnException
+     */
+    private String extractClientId(HttpServletRequest request) throws OAuthClientAuthnException {
+
+        String basicAuthErrorMessage = "Unable to find client id in the request. Invalid Authorization header found.";
+        String authHeader = "Authorization";
+
+        try {
+            Optional<String> signedObject =
```

Comment:
need some inline comment for this method logic

Reply:
Fixed in 1227a3c7eda7627ef2cf6a63f8e54714f18af3a9